Add pandapower interface

jinningwang opened the pull request:
- Add pandapower interface
- Add example of the interface

jinningwang commented:
Hantao,

Let us keep it here. I'll work on it after I finalize the issues we discussed today, expecting by the end of this week.

Regards,
Jinning

jinningwang commented:
Hantao,

I have finalized most of the pandapower interface functions, where the remaining features (enhance the conversion by file I/O, add support of piece-wise cost functions) can be done in the future.

The example I write in ex11 can be used to build the self-test as a baseline.

Regards,
Jinning

jinningwang commented:
The checks failed but I have no idea how to solve them. It seems that the default environment does not include pandapower.

cuihantao commented:
I will work on it next week by adding "optional" packages for ANDES. It will allow installing `interop` (short for interoperability) packages with `pip install andes[interop]` or `pip install -e .[interop]`.

cuihantao commented:
Copied the pandapower interface description to below. We might need to add a bit to it and move it to "Getting Started" - "Interoperability" (a new section.

```
pandapower interface
--------------------

The interface is developed for pandapower 2.7.0, which is not pre-installed in the andes environment.

to_pandapower
.............
This function is used to convert ADNES system (ssa) to pandapower network (ssp). The power flow results
are consistent.

Note:
Line limts are set as 99999.0 in `ssp`.
`SynGen` equipped with `TurbineGov` in `ssa` is set as `controllable=True` in `ssp.gen`.
```

cuihantao commented:
I would like to rename some functions in the pandapower interface. It's difficult to guess what `opp_res` would mean.

cuihantao commented:
I moved `pandapower` from `andes.io` to `andes.interop`. Changed it in the notebook as well. You might need to update your other notebooks @jinningwang

jinningwang commented:
## case300.m issue
Hantao,

Regarding case300, the conclusions are: 1) ANDES and the interface work well, 2) the differences of PF from pandapower are acceptable.

The detailed info is listed below:
Case name:
- Matpower: `case300.m`
- ANDES: `ssa`, converted from `case300.m`
- PP: `ssp`, converted from `ssa`
- PP built-in: `ssp2`, retrieved from pandapower built-in case

PF results comparison:
- `case300.m` vs. `ssa`, consistent
- `ssp` vs. `ssp2`, consistent
- `ssa` vs. `ssp` or `ssp2`, inconsistent
- `case300.m` vs. `ssp` or `ssp2`, inconsistent

Although the PF results are inconsistent, the max gap of voltage amplitude between `ssa` and `ssp` is less than 1e-4.

cuihantao commented:
> ## case300.m issue
> Hantao,
> 
> Regarding case300, the conclusions are: 1) ANDES and the interface work well, 2) the differences of PF between ANDES and pandapower are acceptable.
> 
> The detailed info is listed below: Case name:
> 
> * Matpower: `case300.m`
> * ANDES: `ssa`, converted from `case300.m`
> * PP: `ssp`, converted from `ssa`
> * PP built-in: `ssp2`, retrieved from pandapower built-in case
> 
> PF results comparison:
> 
> * `case300.m` vs. `ssa`, consistent
> * `ssp` vs. `ssp2`, consistent
> * `ssa` vs. `ssp` or `ssp2`, inconsistent
> * `case300.m` vs. `ssp` or `ssp2`, inconsistent
> 
> Although the PF results are inconsistent, the max gap of voltage amplitude between `ssa` and `ssp` is less than 1e-4.

Got it. Thanks!

jinningwang commented:
> Hi @jinningwang , is there further changes to this PR?

All my thoughts have been covered.

cuihantao commented:
Great. We will need a test for the pandapower interface. I might let you do it this time. 

You will find `tests/test_pandapower.py` a barebone. To start, you can move the code from the jupyter notebook to the test script. At the end, you will use

```python
import numpy as np

...

to_pandapower(..., verify=False)

# manually extract the voltage and voltage angle vectors, and do 

np.testing.assert_almost_equal(v_andes, v_pd) 
np.testing.assert_almost_equal(a_andes, v_pd) 
```

to compare the power flow results from ANDES and pandapower. 

One test can be written for `make_link_table`. You can look at the known results for a particular generator and compare the strings.

jinningwang commented:
> Great. We will need a test for the pandapower interface. I might let you do it this time.
> 
> You will find `tests/test_pandapower.py` a barebone. To start, you can move the code from the jupyter notebook to the test script. At the end, you will use
> 
> ```python
> import numpy as np
> 
> ...
> 
> to_pandapower(..., verify=False)
> 
> # manually extract the voltage and voltage angle vectors, and do 
> 
> np.testing.assert_almost_equal(v_andes, v_pd) 
> np.testing.assert_almost_equal(a_andes, v_pd) 
> ```
> 
> to compare the power flow results from ANDES and pandapower.
> 
> One test can be written for `make_link_table`. You can look at the known results for a particular generator and compare the strings.

I would love to do it, a great chance to learn how to write a test.

jinningwang commented:
Hantao,

I added two tests for pandapower interface.
The power flow test cannot pass under default settings. Then I decreased the decimal of ``np.testing.assert_almost_equal`` to 6. Is this acceptable?

Regards,
Jinning

cuihantao commented:
Yes, six decimal digits should be okay.

 

From: Jinning Wang ***@***.***> 
Sent: Friday, March 25, 2022 12:09 PM
To: cuihantao/andes ***@***.***>
Cc: Hantao Cui ***@***.***>; Comment ***@***.***>
Subject: Re: [cuihantao/andes] Add pandapower interface (PR #287)

 

Hantao,

I added two tests for pandapower interface.
The power flow test cannot pass under default settings. Then I decreased the decimal of np.testing.assert_almost_equal to 6. Is this acceptable?

Regards,
Jinning

—
Reply to this email directly, view it on GitHub <https://github.com/cuihantao/andes/pull/287#issuecomment-1079228819> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABSNZA4DNJUGLJ4GVKUFJK3VBXXJXANCNFSM5QGY3X7A> .
You are receiving this because you commented.  <https://github.com/notifications/beacon/ABSNZA4ZQ5E3CSQUT3CIV7LVBXXJXA5CNFSM5QGY3X7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOIBJ3TEY.gif> Message ID: ***@***.*** ***@***.***> >

cuihantao commented:
@jinningwang Jinning, I found a better place for the notebook for pandapower interface. In my latest `develop` branch, I moved it to `docs/examples/pandapower.ipynb`.

In the "Examples" tab, I added a new section for "Interoperability" in the sidebar. It should make navigation more easily. Let me know if you have any thoughts.
